modmail/exts/utils/plugin_manager.py

```diff
@@ -0,0 +1,276 @@
+        r"""
+        Reload extensions given their fully qualified or unqualified names.
+
+        If an extension fails to be reloaded, it will be rolled-back to the prior working state.
+
+        If '\*' is given as the name, all currently loaded extensions will be reloaded.
+        If '\*\*' is given as the name, all extensions, including unloaded ones, will be reloaded.
+        """  # noqa: W605
+        if not plugs:
+            await ctx.send_help(ctx.command)
+            return
+
+        if "**" in plugs:
+            plugs = []
+            for plug, _nul in PLUGINS:
+                plugs.append(plug)
+        elif "*" in plugs:
+            plugs = set(self.bot.extensions.keys()) | set(plugs)
+            plugs.remove("*")
+
+        msg = self.batch_manage(Action.RELOAD, *plugs)
+
+        await ctx.send(msg)
+
+    @extensions_group.command(name="list", aliases=("all",))
+    async def list_command(self, ctx: Context) -> None:
+        """
+        Get a list of all extensions, including their loaded status.
+
+        Grey indicates that the extension is unloaded.
+        Green indicates that the extension is currently loaded.
+        """
+        embed = Embed(colour=Colour.blurple())
+        embed.set_author(
+            name="Plugins List",
+        )
+
+        lines = []
+        categories = self.group_extension_statuses()
+        for category, plugs in sorted(categories.items()):
+            # Treat each category as a single line by concatenating everything.
+            # This ensures the paginator will not cut off a page in the middle of a category.
+            category = category.replace("_", " ").title()
+            plugs = "\n".join(sorted(plugs))
+            lines.append(f"**{category}**\n{plugs}\n")
+
+        log.debug(f"{ctx.author} requested a list of all cogs. Returning a list.")
+        # await Paginator.paginate(lines, ctx, embed, scale_to_size=700, empty=False)
+
+        # since we currently don't have a paginator.
+        output = ""
+        for line in lines:
+            output += line
+        await ctx.send(output)
+
+    def group_extension_statuses(self) -> t.Mapping[str, str]:
+        """Return a mapping of extension names and statuses to their categories."""
+        categories = {}
+        plugs = []
+        for ext, _nul in PLUGINS:
+            plugs.append(ext)
+        for plug in plugs:
+            if plug in self.bot.extensions:
+                status = ":green_circle:"
+            else:
+                status = ":red_circle:"
+
+            path = ext.split(".")
+            if len(path) > BASE_PATH_LEN + 1:
+                category = " - ".join(path[BASE_PATH_LEN:-1])
+            else:
+                category = "uncategorised"
+
+            categories.setdefault(category, []).append(f"{status}  {path[-1]}")
+
+        return categories
+
+    def batch_manage(self, action: Action, *plugs: str) -> str:
```

Comment:
I don't like this duped code. We can probably treat plugins as subclasses of extensions